db/queries.js

```diff
@@ -195,12 +199,139 @@ const deleteS5BarsById = (id) => {
-//     return Promise.reject(new Error('pair name or interval not provided'));
-//   }
+const addMviewByPairName = (pairName, granularity, start = '2 year', withData = false) => {
+  if (!pairName || !granularity) {
+    return Promise.reject(new Error('pair name or granularity not provided'));
+  }
+
+  const withDataSQL = withData ? 'DATA' : 'NO DATA';
+  const mviewName = generateMViewSchemeName(pairName, granularity);
+  const intervalSQL = granularityToSQLString(granularity);
+
+  return getPair({ name: pairName })
+    .then((res) => {
+      if (!res) {
+        throw new Error('pair name does not exist in database');
+      }
+
+      const pairID = res.id;
+
+      return knex.raw(`
+        CREATE MATERIALIZED VIEW ${mviewName}
+        AS
+          WITH intervals AS (
+            SELECT start, start + interval '${intervalSQL}' AS end
+          FROM generate_series(current_date - interval '${start}', current_date, interval '${intervalSQL}') AS start)
+          SELECT DISTINCT
+            intervals.start AS dt,
+            min(bid_l) OVER w AS bid_l,
+            max(bid_h) OVER w AS bid_h,
+            first_value(bid_o) OVER w as bid_o,
+            last_value(bid_c) OVER w as bid_c,
+            sum(bid_v) OVER w AS bid_v,
+            min(ask_l) OVER w AS ask_l,
+            max(ask_h) OVER w AS ask_h,
+            first_value(ask_o) OVER w as ask_o,
+            last_value(ask_c) OVER w as ask_c,
+            sum(ask_v) OVER w AS ask_v,
+            min((bid_l + ask_l) / 2) OVER w AS mid_l,
+            max((bid_h + ask_h) / 2) OVER w AS mid_h,
+            first_value((bid_o + ask_o) / 2) OVER w as mid_o,
+            last_value((bid_c + ask_c) / 2) OVER w as mid_c,
+            sum((bid_v + ask_v) / 2) OVER w AS mid_v,
+            sum(ticks) OVER w AS ticks
+          FROM intervals 
+          JOIN s5bars s5 ON 
+            s5.id_pairs = ${pairID} AND
+            s5.dt >= intervals.start AND
+            s5.dt < intervals.end
+          WINDOW w AS (PARTITION BY intervals.start ORDER BY
+            s5.dt ASC rows BETWEEN UNBOUNDED PRECEDING AND
+            UNBOUNDED FOLLOWING)
+          ORDER BY intervals.start
+        WITH ${withDataSQL}`)
+        .then(() => {
+          return knex.raw(`CREATE UNIQUE INDEX ${mviewName}_index ON ${mviewName} (dt)`);
+        });
+    });
+};
+
+const createMviewCombos = () => {
+  const granularities = ['m1', 'm5', 'm30', 'h1', 'h4', 'd1'];
+
+  return getAllPairs()
+    .then((pairs) => {
+      const pairNames = pairs.map(pair => pair.name);
+
+      const pairGranularityCombos = generatePairGranularityCombos(pairNames, granularities);
+
+      return pairGranularityCombos.reduce((previous, current, index, array) => {
+        return previous.then(() => addMviewByPairName(array[index][0], array[index][1]));
+      }, Promise.resolve());
+    });
+};
+
+const getAllMviews = () => {
+  return knex.select(knex.raw('oid::regclass::text')).from('pg_class').where('relkind', 'm');
```

Comment:
Probably want to put each function you're chaining on a new line, like you do below